epidose/device/wps_scanner_d.sh

```diff
@@ -0,0 +1,123 @@
+
+# Pick up utility functions relative to the script's source code
+UTIL="$(dirname "$0")/util.sh"
+
+# Source common functionality (logging, WiFi)
+# shellcheck source=epidose/device/util.sh
+SERVER_URL="http://ha-server:5010" . "$UTIL"
+
+# Light up LED lights to indicate network status messages.
+# precondition: WiFi should be turned on.
+# Returns an exit code of 0 if a network connection is established,
+# while exit code 1 denotes a failed attempt to connect to a WPS network,
+# exit code 2 shows that there are no available WPS networks in the area,
+# and value 3 refers to an invalid LED color option.
+wps_led_blink()
+{
+  run_python device_io --"$1"-on
+  case "$1" in
+	"green") log "Obtained network connection"; return 0;;
+	"red") log "Failed to connect to a network"; return 1;;
+	"orange") log "No WPS network found in the area"; return 2;;
+	*) log "The given LED option is not available"return 3;;
+  esac
+  sleep 3
+  run_python device_io --"$1"-off
+}
+
+# Connect to a know network if available,
+# otherwise try to connect to the network with
+# the stongest signal by using WPS.
+# precondition: none
+# Returns exit codes based on the return codes
+# of the wps_led_blink function
+wps_connect()
+{
+  network_status=$(ifconfig wlan0 | grep "inet ")
+
+  # Check if device is connected to any network
+  if [ -z "$network_status" ] ; then
+    log "Iniating WPS connection"
+ 
+    # Suspend watchdog's LED activities
+    touch "$SUSPEND_WATCHDOG_LED"
+  
+    # Get WPS network with the strongest signal
+    get_network_with_strongest_signal=$(wpa_cli -i wlan0 scan_results | grep WPS | sort -r -k3 | tail -1)
+    if [ ! -z "$get_network_with_strongest_signal" ] ; then
+      # Get bssid of the strongest network signal and connect to it 
+      echo "$get_network_with_strongest_signal" | awk '{print $1}' | xargs wpa_cli -i wlan0 wps_pbc
+     
+      # Get network name
+      network_name=$(echo "$get_network_with_strongest_signal" | awk '{print $NF}')
+      log "Trying to connect to $network_name"
+    
+      # Obtain dynamic IP address
+      dhclient wlan0
+
+      get_ip_address=$(ifconfig wlan0 | grep "inet ")
+    
+      # Check whether the connection is established and light up the LED accordingly
+      if [ -z "$get_ip_address" ] ; then
+        # If failed to connect to a network, turn on red LED light
+	wps_led_blink red
+	exit_code=$?
+      else
+        # If connected to a network, turn on green LED light
+	wps_led_blink green
+	exit_code=$?
+      fi    
+    else
+      # If there is no available WPS network in the area. turn on orange LED light	    
+      wps_led_blink orange
+      exit_code=$?
+    fi
+  
+    # Release watchdog
+    rm "$SUSPEND_WATCHDOG_LED"
+    
+    return "$exit_code"
+  fi
+
+  return 0
+}
+
+while : ; do
+  # Wait for the Wifi button to be pressed
+  run_python device_io --wifi-wait
+  wifi_acquire
+  if wps_connect; then
+    validity_time=$(get_filter_validity_age)
```

Comment:
Is this required?  Won't the other daemon handle it?  Can't it create a race condition?  Consider just notifying the other daemon, if needed  e.g. by killing its sleep process, which can be recorded in `/var/run/epidose` with something like `sleep … & ; echo … ; wait …`

Reply:
If I have understood correctly, you are suggesting to make the following in the update_filter_d

`sleep "$validity_time" & sleep_pid=$!; echo $sleep_pid > /var/run/epidose_update_filter_sleep ; wait $sleep_pid`

and the wps_scanner_d should kill the sleep_pid after it updates the filter?

`get_new_filter
 kill -9 $(cat /var/run/epidose_update_filter_sleep)
`
Another question is if the wps_scanner should update the cuckoo filter after it expires
or anytime it obtains a network connection?

Reply:
1. Exactly!
2. Good point. The cost of getting it is small, so let's get it while it can.

Reply:
Ok, done. I have submitted a new commit.